### DISCUSSION TITLE
How to better custom our own image?

### DISCUSSION THREAD

**Discussion by lam0819** [serversideup/docker-php#181] [Q&A] [answered]
I want to make the deployment process easier, so I plan to build everything in a single image and run the queue and schedule from it. To achieve this, I have customized my own image with additional code on s6, as shown in the screenshots below.

For the queue, I have included a script that checks if an Artisan file exists and assumes that Laravel is configured if it does. If AUTORUN_ENABLED is set to 'true' and AUTORUN_LARAVEL_QUEUE is set to 'true', the script runs the Laravel queue with specific parameters. If Laravel is not detected or is disabled, the script skips running the queue.

I am unsure if this is the correct approach for customization, so please let me know if there are any recommendations for improvement.

My Dockerfile
![carbon (9)](https://github.com/serversideup/docker-php/assets/68211972/23bac0fe-577c-4416-8ae8-f9cf8102aa8c)

S6 stuff
![Screenshot 2023-05-25 at 10 42 10 AM](https://github.com/serversideup/docker-php/assets/68211972/09b36d4a-42b0-4bd7-80f5-25862c725d54)

for qaueue
```
#!/command/with-contenv bash

# Exit on error
set -e

# Check to see if an Artisan file exists and assume it means Laravel is configured.
if [ -f $WEBUSER_HOME/artisan ] && [ ${AUTORUN_ENABLED:="true"} == "true" ]; then
        echo "🏃‍♂️ Checking for Laravel Queue default ..."

        ############################################################################
        # Automated AUTORUN_LARAVEL_QUEUE
        ############################################################################
        if [ ${AUTORUN_LARAVEL_QUEUE:="false"} == "true" ]; then
            echo "🚀 Running queue default..."
            s6-setuidgid webuser php $WEBUSER_HOME/artisan queue:work --sleep=3 --tries=3 --daemon --timeout=0
        fi

else
    echo "👉 Skipping Laravel Queue because we could not detect a Laravel install or it was specifically disabled..."
fi

exit 0
```

**Comment by jaydrogers** [comment, marked as answer]
Great stuff!

Few quick comments:

### Reduce the copy command
<img width="727" alt="image" src="https://github.com/serversideup/docker-php/assets/3174134/12c851d2-0e87-40b0-a4a7-daf394bb58cb">

Minimize your layers as much as possible. Rather than copying each file, you can copy the entire directory (and its sub directories & files) all at once. That's what I do here:

https://github.com/serversideup/docker-php/blob/69641f3d13abde0212c02403c6b2c0fe599a805c/src/fpm/Dockerfile#L44

### Running a queue & scheduler
I see you're injecting these scripts to run the services. Although that may work, you lose the supervision of the process.

Instead, consider separating out your processes into multiple containers -- so you would have:

1. PHP
2. Scheduler
3. Queue

See these examples:
https://serversideup.net/open-source/docker-php/docs/laravel/laravel-task-scheduler
https://serversideup.net/open-source/docker-php/docs/laravel/laravel-queue

Hope this helps!

**Reply by jaydrogers** [reply]
One other thing to add:

### Setting file permissions
<img width="337" alt="image" src="https://github.com/serversideup/docker-php/assets/3174134/801d3229-fde7-4fed-baf1-ade48a3da4a4">

You don't need to run this command. You can include `--chown=9999:9999` if you're using the default GID and UID with your `COPY` command (like below)

https://github.com/serversideup/financial-freedom/blob/10dcc4f60637874031a8fc80f3a092cd0a943c11/Dockerfile#L5
